[extension/ecsobserver]  Add get ip and port from task

pingleig opened the pull request:
**Description:** 

Follow up on #2734  Add get private ip and mapped port from ecs task

**Link to tracking Issue:**

- Feature request #1395 

Previous PRs

- Target and Task definition (merged) https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2939
- Config PR (merged) https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2377
- Original PR (closed) #2734 

**Testing:** 

unit test

**Documentation:** 

No new doc. See [existing doc](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/ecsobserver/README.md)

pingleig committed:
ext: ecsobserver Add get ip and port from task

pingleig committed:
ext: ecsobserver Add merge container and target

pingleig commented:
cc @jrcamp @mxiamxia

pingleig commented on extension/observer/ecsobserver/task.go:
```diff
@@ -65,3 +80,67 @@ func (t *Task) ContainerLabels(containerIndex int) map[string]string {
+	mode := aws.StringValue(t.Definition.NetworkMode)
+	switch mode {
+	case ecs.NetworkModeNone:
+		return 0, fmt.Errorf("task has none network mode %s", arn)
```
For the errors, I plan to define error structs later to keep the PR smaller. They will change to dedicated error types for the telemetry https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/observer/ecsobserver#metrics

pingleig replied:
Changed to error struct. Might add an interface to indicates if this a fatal error. Though the error handling logic for the extension is basically log and ignore, because you can have invalid task/extension config while still getting some valid targets. Based on experience form existing customers (in cloudwatch agent), they don't want to stop the collection when some invalid task shows up (e.g. write a docker label `ECS_PROM_PORT: 123abc`).

The plan is to export those errors are metrics https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/observer/ecsobserver#metrics and if user care about those errors, they can set alarm on those metrics.

Or we can have a strict mode or development mode to exit on error. It's hard to exit on error as well, it seems there is no good way for extension to tell the service it has crashed because [Start](https://github.com/open-telemetry/opentelemetry-collector/blob/main/component/component.go#L46) starts the extension in background.

pingleig commented on extension/observer/ecsobserver/task.go:
```diff
@@ -65,3 +80,67 @@ func (t *Task) ContainerLabels(containerIndex int) map[string]string {
+	// Default network mode is bridge on EC2
+	case "", ecs.NetworkModeHost, ecs.NetworkModeBridge:
+		if t.EC2 == nil {
+			return "", fmt.Errorf("task has no network mode and no ec2 info %s", arn)
```
No, they should be in same case, when using ec2, as long as you are not using awspc, the task ip is host ip. Though I might need to double check fargate behaviour, fargate always uses awsvpc but I am not sure if the API accepts `""` as network mode (it does for ec2).

pingleig replied:
Checked fargate, it has to be awsvpc https://docs.aws.amazon.com/AmazonECS/latest/userguide/fargate-task-defs.html#fargate-tasks-networkmode. So `""` means it must be ECS EC2 bridge, otherwise it will get rejected by ECS API and we won't see it the extension.

pingleig commented on extension/observer/ecsobserver/task.go:
```diff
@@ -65,3 +80,67 @@ func (t *Task) ContainerLabels(containerIndex int) map[string]string {
+				return aws.Int64Value(v.HostPort), nil
+			}
+		}
+		return 0, fmt.Errorf("port %d not found for network mode %s", containerPort, mode)
```
I think we can just switch to structured error to include info about both the rule and task.

pingleig replied:
Add structured error and actually it saves a lot of fmt.Errrorf since all the network mode errors are same except for the mode part ...

pingleig commented on extension/observer/ecsobserver/matcher.go:
```diff
@@ -57,6 +57,24 @@ type MatchedContainer struct {
+// If there are duplicated 'key' we honor existing target and does not override.
+// The duplication could happen if there are several rules matching same target.
+func (mc *MatchedContainer) MergeTargets(newTargets []MatchedTarget) {
+	for _, newt := range newTargets {
+		for _, old := range mc.Targets {
+			// If port and metrics_path are same, then we treat them as same target and keep the existing one
+			if old.Port == newt.Port && old.MetricsPath == newt.MetricsPath {
+				continue
+			}
+			mc.Targets = append(mc.Targets, newt)
+		}
+	}
```
good catch,  it's not. I think we can change the label to `NextTarget` I don't recall if there are style guide for all caps goto labels. 

I will update the unit test as well.  

Those logic actually rarely got trigged in our [e2e test](https://github.com/aws-observability/aws-otel-test-framework) because most of sample workload config does not match same task using different matching rules and assign different port path combination. But since we don't allow multiple metrics path in one config, this logic is one way to workaround it. (btw: It's different from k8s discovery where all the ports are discovered and filtered later in relabel)

```
- ports:
    - 123
    - 456
  path: /metrics
- ports:
   - 789
   path: /another-metrics-path
```

The reason we don't allow config both is it's a bit confusing and try to be close to existing cloudwatch agent's ecs prometheus discovery config.

```
If both ports and paths are array

- ports:
   - 123
   - 456
   paths:
  - /metrics
  - /another-metrics

Will result in the following, which is not a common case for workloads in cloudwatch's container insight. 
But if it is common for other workloads, we may switch to this config pattern.

:123/metrics
:123/another-metrics
:456/metrics
:456/another-metrics
```

mxiamxia replied:
The discovery ticker adds or merges the newly matched targets with the old/cached targets. Do we need to remove the old target from Task structure it is not matched anymore?

pingleig replied:
We don't cache task, we only (LRU) cache task definition, ec2 instance etc. We don't save previous matched result in memory. Every tick we simply scan all the tasks, do the match and replace the content of entire file sd. Prometheus itself (discovery manger) will handle the diff logic and the discovery plugins can just do full dump. And there is no way for plugin like file sd to communicate with prometheus.

pingleig replied:
Fixed in latest commit, also added some unit test.

tigrannajaryan commented:
Please make sure the build passes.

pingleig commented:
I think the build failed (12days ago due to dockerhub problems). I need to push new commits to address review comments so we don't need to retry the build manually (I think retry on circleci has permission requirement.)

[circleci logs](https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector-contrib/12460/workflows/86511be7-a5ac-4726-add7-b50aeb724821/jobs/107050)

```
Failed
    TestDefaultMetricsIntegration: container.go:109: failed pulling image docker.io/library/nginx:1.17: Error response from daemon: received unexpected HTTP status: 503 Service Unavailable
```

pingleig committed:
ext: ecsobserver Fix MergeTarget logic

Based on review comments from Anthony

- Add error struct for ip and port error, we can report metrics later.
- MergeTarget is wrong, add unit test to make sure it is merging new
  target only once instead of multiply by existing targets :D
- Fix typos

pingleig commented on extension/observer/ecsobserver/matcher.go:
```diff
@@ -57,6 +57,24 @@ type MatchedContainer struct {
+// MergeTargets adds new targets to the set, the 'key' is port + metrics path.
+// The 'key' does not have ip because all the targets from on container has same ip.
+// If there are duplicated 'key' we honor existing target and does not override.
+// The duplication could happen if there are several rules matching same target.
```
Fixed (copy and pasted) in latest commit.

pingleig commented on extension/observer/ecsobserver/task.go:
```diff
@@ -65,3 +80,67 @@ func (t *Task) ContainerLabels(containerIndex int) map[string]string {
+}
+
+// MappedPort returns 'external' port based on network mode.
+// EC2 bridge gets random host port while EC2 host/awsvpc uses whatever specified by user.
```
Fixed in latest commit.

pingleig commented on extension/observer/ecsobserver/task.go:
```diff
@@ -65,3 +80,67 @@ func (t *Task) ContainerLabels(containerIndex int) map[string]string {
+				}
+			}
+		}
+		return 0, fmt.Errorf("port %d not found for network mode %s", containerPort, mode)
```
Changed to error struct as well.

pingleig commented:
@Aneurysm9 @mxiamxia PTAL

btw: build failed for [load test](https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector-contrib/13005/workflows/9f3efa6c-8af8-4d91-bdbf-5c6fa4d4db5a) the extension is not enabled, so it should be loadtest's own problem.

mxiamxia reviewed (approved):
Thanks!
